core/kernel/user_access.c

```diff
@@ -74,6 +94,41 @@ TEE_Result copy_to_user_private(void *uaddr, const void *kaddr, size_t len)
 	return res;
 }
 
+void *bb_alloc(size_t len)
+{
+	struct user_mode_ctx *uctx = get_current_uctx();
+	size_t offs = 0;
+	void *bb = NULL;
+
+	if (uctx && !ADD_OVERFLOW(uctx->bbuf_offs, len, &offs) &&
+	    offs <= uctx->bbuf_size) {
+		bb = uctx->bbuf + uctx->bbuf_offs;
+		uctx->bbuf_offs = ROUNDUP(offs, sizeof(long) * 2);
+	}
+	return bb;
+}
+
+void bb_free(void *bb, size_t len)
+{
+	struct user_mode_ctx *uctx = get_current_uctx();
+	size_t offs = ROUNDUP((vaddr_t)bb + len, sizeof(long) * 2);
```

Comment:
`off = ROUNDUP((vaddr_t)bb + len, sizeof(long) * 2) - uctx->bbuf;` ?

Reply:
You're right. I'll rewrite the function to make it easier to follow.